### PR TITLE
storage: test evaluateBatch

### PR DESCRIPTION
**Pull request by tbg**
`evaluateBatch` didn't have any direct testing, though it contains
enough logic that could use some.

Since I'll be modifying it shortly (to add TargetBytes support) I went
ahead and added comprehensive unit testing around the MaxSpanRequestKeys
pagination.

I also made a small (noop) change to `evaluateBatch` itself in
preparation for an upcoming change in the MVCC layer to treat
`maxKeys==0` as "unbounded", which will allow us to take care
of a long-standing TODO to move maxKeys into `MVCCScanOptions`.
I'll leave actually doing that for the next PR, though.

Release note: None

### CODE REVIEW

**Review by nvb — COMMENTED** [review]
Nice tests.

I've been recently thinking about creating datadriven tests for `storage/batcheval`. That wouldn't quite capture what you're trying to test here, but it would complement it. I'm curious if the idea has also crossed your mind.

> Reviewed 8 of 8 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-:-M-vYfhlB599o-jvVhnT:b-xlal9z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_evaluate_test.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vTFF12bWnzYUwvnFR:-M-vTFF234W_TkQUOjzp:bj4sgoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L33)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	var tcs []testCase
> ```

nit: why the appending instead of the standard slice literal, which allows for implicit type 
___
*[pkg/storage/replica_evaluate_test.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vUVG16YHmvNxxbz-H:-M-vUVG16YHmvNxxbz-I:begyeyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L60)):*
> ```Go
> 		},
> 	}, testCase{
> 		// Ditto in reversej
> ```

"reversej"
___
*[pkg/storage/replica_evaluate_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vWY7W1wcuzjcOizgr:-M-vWY7W1wcuzjcOizgs:bxhblai) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L175)):*
> ```Go
> 			writeABCDEF(t, d)
> 			d.ba.Add(scan("a", "c"))
> 			d.ba.Add(get("e"))
> ```

Huh, this is surprising. I think this relates to the large comment before, but I don't directly see the link (or maybe I'm just not fully understanding it). Why doesn't the `get` return a response? Do the requests get sorted by start key at some point?
___
*[pkg/storage/replica_evaluate_test.go, line 206 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vXlwI-B9PL6kZxyIz:-M-vXlwI-B9PL6kZxyJ-:bbvzxs7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L206)):*
> ```
> strings.Replace(tc.name, " ", "_", -1)
> ```

Isn't this already done for you?
___
*[pkg/storage/replica_evaluate_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vY0Ou3dc4H8RRKzDS:-M-vY0Ou3dc4H8RRKzDT:b89zvwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L231)):*
> ```Go
> }
> 
> func runEvalBatchTest(t *testing.T, tc testCase) {
> ```

nit: any reason not to inline a lot of this (mainly `testCase` and `runEvalBatchtest`) in the test function?
___
*[pkg/storage/replica_evaluate_test.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vYOZ03SFZ2VMutrL6:-M-vYOZ03SFZ2VMutrL7:b-kif6z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L267)):*
> ```Go
> }
> 
> func scan(s, e string) *roachpb.ScanRequest {
> ```

Don't we already have functions like these in `pkg/storage` tests? I'd be anxious about leaking these functions into the package namespace.
___
*[pkg/storage/batcheval/eval_context.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vQWEDEQ2aquFVDzPE:-M-vQWEDEQ2aquFVDzPF:b-b3cyki) ([raw file](https://github.com/cockroachdb/cockroach/blob/173d49617ae29d6f8001f484267169ab6142172b/pkg/storage/batcheval/eval_context.go#L116)):*
> ```Go
> }
> 
> // Impl returns the MockEvalCtx as an EvalContext. Modifications to the
> ```

Is `Impl()` the correct name for this? Isn't it returning the opposite (i.e. the implementation as an interface, not the interface's implementation)?

**Review by tbg — COMMENTED** [review]
>  I'm curious if the idea has also crossed your mind.

Definitely! I thought about going down that road for the tests in this PR but then I realized there were too many questions about what that would look like and I got worried that I would just load myself up with a week worth of work that wouldn't get me closer to getting these byte limits hooked up. Happy to chat about what these could look like. On the whole though, I would be slightly more interested in testing more of the layers above `evaluateBatch`. It has several layers of callers sitting on `(*Replica)` which seem to have very little exposure to direct tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-:-M-xeN4-4EVODEgrrEgX:b-4bfio3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_evaluate_test.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vTFF12bWnzYUwvnFR:-M-xfG4s-DyXzcbCzWux:b-xbo99q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: why the appending instead of the standard slice literal, which allows for implicit type 
</blockquote></details>

Artifact of earlier iteration of the test in which I had all of these helper functions defined locally. I really dislike how unreadable tests get with all of that, so I was hoping to put the methods close to the test case that uses them instead. That was still noisier than I could stomach, so I put the helpers into the package namespace, but didn't go back to simplify the testCase definition. Done now.
___
*[pkg/storage/replica_evaluate_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vWY7W1wcuzjcOizgr:-M-xg6DcDj9Gx2LTQ16i:b-5miist) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Huh, this is surprising. I think this relates to the large comment before, but I don't directly see the link (or maybe I'm just not fully understanding it). Why doesn't the `get` return a response? Do the requests get sorted by start key at some point?
</blockquote></details>

Poor writing on my part. The `nil` ignored the response (i.e. checks nothing about it). This was silly because it's straightforward to treat the Get as a Scan, so I did that instead. Values are returned and the test now reflects that.
___
*[pkg/storage/replica_evaluate_test.go, line 206 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vXlwI-B9PL6kZxyIz:-M-xi7sd5Ncgm-MKBimV:b5r3pq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> strings.Replace(tc.name, " ", "_", -1)
> ```

Isn't this already done for you?
</blockquote></details>

Oh, nice.
___
*[pkg/storage/replica_evaluate_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vY0Ou3dc4H8RRKzDS:-M-xk0my23R1Zchn5Uyz:bgsk5by) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L231)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: any reason not to inline a lot of this (mainly `testCase` and `runEvalBatchtest`) in the test function?
</blockquote></details>

I did inline `runEvalBatchTest`, there wasn't a good reason not to (I was initially thinking about the TargetBytes versions of these tests, where maybe we'd need to extend this test somewhat, but I'll see when I get there). See the discussion below for `testCase` - if I inline that I have to inline all.
___
*[pkg/storage/replica_evaluate_test.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vYOZ03SFZ2VMutrL6:-M-xiha65zkNKCSHyQsg:bf8htec) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L267)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we already have functions like these in `pkg/storage` tests? I'd be anxious about leaking these functions into the package namespace.
</blockquote></details>

I'm not in love with either option. I feel fairly strongly that locally defining the myriad of helper types and methods into the test body itself makes these tests a lot less effective to later revisit, especially since many of them necessarily have to precede the definition of the test cases. I've also found in many tests I see (and wrote in the past) that defining helpers locally tends to have them close over variables from the surrounding scope, which is sometimes fine but goes out of whack quickly. On the other hand, I do wish we could scope them "local to the test" somehow.

To answer your concrete question, we have these two:

```
pkg/storage/replica_test.go:func scanArgs(start, end []byte) roachpb.ScanRequest {
pkg/storage/replica_test.go:func reverseScanArgs(start, end []byte) roachpb.ReverseScanRequest {
```

Ultimately I found that concerns around leaking into the package namespace were more theoretical (so what? If we conflict we can move things) and getting the test to be as clean as possible seemed more important. Let me know if you disagree strongly enough to want to inline everything, in which case I'll just do that.
___
*[pkg/storage/batcheval/eval_context.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vQWEDEQ2aquFVDzPE:-M-xl6aL8QXqI7JoqUOC:b-8ovq1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/173d49617ae29d6f8001f484267169ab6142172b/pkg/storage/batcheval/eval_context.go#L116)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is `Impl()` the correct name for this? Isn't it returning the opposite (i.e. the implementation as an interface, not the interface's implementation)?
</blockquote></details>

True. I renamed to `EvalContext()`.

**Review by nvb — COMMENTED** [review]
> I thought about going down that road for the tests in this PR but then I realized there were too many questions about what that would look like and I got worried that I would just load myself up with a week worth of work that wouldn't get me closer to getting these byte limits hooked up.

Hah, I came to the same conclusion. Specifically, I was interested in testing the LocalResult returned by command evaluation but realized that there was a lot more that someone might want to test than just that.

> Reviewed 10 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-:-M-zSa_1AuMAhoO-CC82:b-uq8zo1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/storage/replica_evaluate_test.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-M-vYOZ03SFZ2VMutrL6:-M-zQb9n80NLglP6N3I1:br5sukd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17350df233e5a58789f9b9193edea459f6131ab/pkg/storage/replica_evaluate_test.go#L267)):*
> I feel fairly strongly that locally defining the myriad of helper types and methods into the test body itself makes these tests a lot less effective to later revisit, especially since many of them necessarily have to precede the definition of the test cases. I've also found in many tests I see (and wrote in the past) that defining helpers locally tends to have them close over variables from the surrounding scope, which is sometimes fine but goes out of whack quickly.

Yes, I agree with you on all of this.

> Let me know if you disagree strongly enough to want to inline everything, in which case I'll just do that.

I don't want to inline anything. I'm more interested in using these existing `getArgs`/`scanArgs`/`reverseScanArgs` functions instead of defining new ones. I'd hate to have to ask the question everything time I'm writing a new test in `pkg/storage` whether I should be using `scanArgs` or `scan` and what the difference between them is in the first place.

Was the issue that those functions all return values instead of pointers (and by extension, `roachpb.Request`s instances) or was the issue that they take byte slices instead of strings?

I think the first could be solved by switching all of those functions over to return pointers. That's probably a good change on its own given that we then always need to take the return value's address to use it and that all of these types have `ShallowCopy` methods already in case any caller is relying on value semantics.

The second issue could be solved by adding `getArgsString`/`scanArgsString`/`reverseScanArgsString` functions that sit right next to the existing functions.

The TL;DR is that I'm primarily concerned with fragmenting the helper function namespace unnecessarily.

**Comment by tbg** [comment]
I'll get back to your comments tomorrow (basically sgtm), but other than
those is this LGTM? I'd like to get this in since I have another PR on top
of it.

**Comment by nvanbenschoten** [comment]
Yes, other than that, everything else LGTM.

**Review by tbg — COMMENTED** [review]
:thumbsup: TFTR! I colocated and cleaned up the helpers.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45008#-:-M0qQvTaCt6RKbM3j-vP:b-1pve6v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

**Comment by tbg** [comment]
bors r=nvanbenschoten
